### DISCUSSION TITLE
A little more documentation please

### DISCUSSION THREAD

**Discussion by DGilmour5** [AMalininHere/camunda-worker-dotnet#46] [Q&A]
Greetings,
I'm really interested in your work and supporting your efforts if I can.
I'm a little confused on some of the changes you've implemented since 0.10. Can you provide some comments in the source code or more in the ReadMe?
For instance.. 

- [x] What are the different methods doing in the worker builder? I get AddHandler, but what are the others doing?
- [x] What is the impact of the worker count? 
- [x] What's the proper use of the [HandlerVariables] attribute? 
- [x] How can I get all of the variables?
- [x] How should I deal with External task topics that I don't have a specific handler for? Can I have some kind of catch all?
- [x] What's the proper use of the CompleteTask object?
- [x] What configuration do I need if I'm running multiple instances of my client code? Unique worker names?

Again.. really thankful for your work. When do you think you'll get to version 1.0? And when will the breaking changes slow down?

**Comment by TechnoBerry** [comment]
Hello!

Thank you for your feedback!

When I release new version of the library I try to describe important breaking changes in [release notes](https://github.com/AMalininHere/camunda-worker-dotnet/releases). I understand the importance of having documentation, but I have no experience in providing it. Adding documentation to the library is one of the most important tasks for the 1.0 release. Thank you for the list of issues that need to be covered in the documentation.

As for breaking changes, unfortunately, during the initial design, I did not consider the case of multi-tenancy (#16), the implementation of which I am currently working on. The implementation of this feature requires major changes in the library.  I think that after the next version (most likely 0.14.0), the codebase will be stabilized and the path to 1.0.0 will become clearer.

**Comment by DGilmour5** [comment]
Well, maybe I'll chip in and help a bit as I dig into the internals. I'll fork the project and see if I can contribute in meaningful ways.
If you wouldn't mind taking a few moments to answer the questions quickly, I'm sure it'll be helpful for me and others.

**Comment by TechnoBerry** [comment]
I will be glad of your help, but first I have to make a backlog so that your work is not in vain. As for your questions, some questions relate to things that I plan to change in the next version and I can only outline a general vision. Here are the answers for some of questions. I will prepare answers to the rest later.

> What is the impact of the worker count?

This is the number of [parallel workers](https://github.com/AMalininHere/camunda-worker-dotnet/blob/master/src/Camunda.Worker/Execution/DefaultCamundaWorker.cs) with same configuration. Every worker performs ["fetch and lock -> process"](https://docs.camunda.org/manual/7.16/user-guide/ext-client/#bootstrapping-the-client) routine in infinite loop (see [RunAsync](https://github.com/AMalininHere/camunda-worker-dotnet/blob/master/src/Camunda.Worker/Execution/DefaultCamundaWorker.cs#L43-L65)). [The hosted service](https://github.com/AMalininHere/camunda-worker-dotnet/blob/master/src/Camunda.Worker/Execution/WorkerHostedService.cs) is responsible for starting the workers.
In the [sample app](https://github.com/AMalininHere/camunda-worker-dotnet/blob/master/samples/SampleCamundaWorker), the number of parallel workers is `1`. This means that only one worker performs "fetch and lock -> process" routine.

> What's the proper use of the CompleteTask object?

Did you mean [CompleteResult](https://github.com/AMalininHere/camunda-worker-dotnet/blob/master/src/Camunda.Worker/CompleteResult.cs)? If so, then this is an analogue of OkObjectResult from ASP.NET Core. Instances of this class represent successful processing of an external task. As you can see [here](https://github.com/AMalininHere/camunda-worker-dotnet/blob/master/src/Camunda.Worker/CompleteResult.cs#L27), processing such instances internally causes camunda to [complete the task](https://docs.camunda.org/manual/7.16/reference/rest/external-task/post-complete/).

> What configuration do I need if I'm running multiple instances of my client code? Unique worker names?

Unfortunately, I don't have an answer to this question. It seems to me that this generally applies to interaction with camunda. This question is relevant not only for this library, but also for Java and JS libraries. You can try to find the answer in their forum. (https://forum.camunda.org/)

**Comment by TechnoBerry** [comment]
> What's the proper use of the [HandlerVariables] attribute?

This attribute is used to tell camunda which variables the handler needs. Sometimes it might be a local variables. You can just use it to "declare API" of your handler.

> How can I get all of the variables?

The documentation of ["fetch and" lock"](https://docs.camunda.org/manual/7.16/reference/rest/external-task/fetch/#request-body) says "If not provided - all variables will be fetched." So if you don't use the [HandlerVariables] attribute, you will get all the variables.

> How should I deal with External task topics that I don't have a specific handler for? Can I have some kind of catch all?

Set of topic names is required to perform ["fetch and lock"](https://docs.camunda.org/manual/7.16/reference/rest/external-task/fetch/). Your program doesn't receive external task with topic name that your program doesn't request.

**Comment by TechnoBerry** [comment]
> What are the different methods doing in the worker builder? I get AddHandler, but what are the others doing?

At the moment there are 4 methods that can be used to modify the work of the worker:
- `AddTopicsProvider` is designed to convert information about registered handlers into topics that will be used in "fetch and lock" requests. There is a task (#43) to change this abstraction.
- `AddEndpointProvider` is designed to change the mechanism for selecting a handler based on information about an external task. I plan to change the signature of this abstraction.
- `ConfigurePipeline` #2
- `ConfigureEvents` #4
